### PR TITLE
feat: [FC-1] load more message

### PR DESCRIPTION
**Pull request by huydosgtech**

### CODE REVIEW

**Comment by huydosgtech** [comment]
Video Demo:

https://github.com/saigontechnology/rn-firebase-chat/assets/105196859/ffded200-191c-4c7e-81ab-81c8885f3685

**Comment by honghoangsts** [comment]
@huydosgtech Please remove the code not relative to load more feature.

**Comment by huydosgtech** [comment]
> @huydosgtech Please remove the code not relative to load more feature.

I updated. Please check

File: src/chat/ChatScreen.tsx [hunk 1]
```diff
@@ -96,6 +98,16 @@ export const ChatScreen: React.FC<ChatScreenProps> = ({
     [firebaseInstance, memberIds, partners]
   );
 
+  const onLoadEarlier = useCallback(async () => {
+    if (conversationRef.current?.id) {
+      const res = await firebaseInstance.getMoreMessage();
+      setHasMoreMessages(res.length > 0);
```
Comment (honghoangsts) [inline src/chat/ChatScreen.tsx:104]: Why has more message when res.length > 0?

**Review by honghoangsts — CHANGES_REQUESTED** [review]

File: src/chat/ChatScreen.tsx (hunk 1 above)
Comment (huydosgtech) [reply src/chat/ChatScreen.tsx:104]: because we need disable loadEarlier={hasMoreMessages}, if not loading indicator will show forever

Comment (honghoangsts) [reply src/chat/ChatScreen.tsx:104]: if you call `getMoreMessage` and it return only 1 message, should we show load more?

Comment (huydosgtech) [reply src/chat/ChatScreen.tsx:104]: No, **onLoadEarlier** function only called when message res.length >= 20 at line 65, So when go to the first time will not show any load more. Only when more then 20 messages and scroll to top, **onLoadEarlier** will trigger and load more

Comment (honghoangsts) [reply src/chat/ChatScreen.tsx:104]: what happen if we have 21 messages?

Comment (huydosgtech) [reply src/chat/ChatScreen.tsx:104]: **hasMoreMessages** is true and **onLoadEarlier** depend on **loadEarlier** function is true. So it still working well. This is video demo when 21 messages:

https://github.com/saigontechnology/rn-firebase-chat/assets/105196859/139f6de0-d859-4d32-9928-1b10f6b24078

Comment (huydosgtech) [reply src/chat/ChatScreen.tsx]: I addressed check duplicate, please help me verify it again. Thank you

File: src/chat/ChatScreen.tsx [hunk 2]
```diff
@@ -96,6 +99,24 @@ export const ChatScreen: React.FC<ChatScreenProps> = ({
...
+      const res = await firebaseInstance.getMoreMessage();
+      setHasMoreMessages(res.length === 20); // if the length of the response is 20, it means there are more messages to load
+      setMessagesList((previousMessages) =>
+        GiftedChat.prepend(previousMessages, res)
+      );
+    }
+
+    setTimeout(() => {
+      isLoadingRef.current = false;
+    }, 300);
```
Comment (honghoangsts) [inline src/chat/ChatScreen.tsx:117]: why we need delay 300ms?

Comment (huydosgtech) [reply src/chat/ChatScreen.tsx:117]: yes, this need timeout too make sure not duplicated **onLoadEarlier** function in 300ms

**Review by honghoangsts — APPROVED** [review]

**Pull request merged**